P003 Pulse Microseconds

clumsy-stefan opened the pull request:
Use micros() for time measurement instead of milis(). Like this the Pulse counter "Time" value uses more precision and also frequencies above 100Hz (>eg. Fans with more than 6'000 RPM) can be measured "instantly" without any processing overhead. With milis() steps are 6'000, 6'666, 7'500. With micros() it gives quite an accurate instant RPM reading also below 10ms per pulse.

Unfortunately this breaks backward compatibility for the "Time" and "Debounce" values as they are now in microseconds.

clumsy-stefan committed:
P003 Pulse Microseconds

Use micros() for time measurement instead of milis(). Like this the Pulse counter "Time" value uses more precision and also frequencies above 100Hz (>eg. Fans with more than 6'000 RPM) can be measured "instantly" without any processing overhead. With milis() steps are 6'000, 6'666, 7'500. With micros() it gives quite an accurate instant RPM reading also below 10ms per pulse.

Unfortunately this breaks backward compatibility for the "Time" and "Debounce" values as they are now in microseconds.

TD-er commented:
> Unfortunately this breaks backward compatibility for the "Time" and "Debounce" values as they are now in microseconds.

Why not divide the values by `1000.0f` ?
Then you still have the same unit of measure, only with a higher resolution.

clumsy-stefan commented:
> Why not divide the values by `1000.0f` ?
> Then you still have the same unit of measure, only with a higher resolution.

in fact, you would need to have one multiplication by 1000 for the debaonce and one division to store the new value in the UserVar.
I also had this idea, but then as soon as a (fast) signal was on a pin while booting it crashed immerdiately. Not completely sure why, but I guess the additional calculations within the interrupt caused too much delay. It seemed to work at low frequencies, but as soon as I increased the fan-speed the unit crashed again..

TD-er commented:
All code (and variables) accessed in an interrupt callback function should be declared volatile and also marked with ICACHE_ATTR.
So the interrupt functions should store it first in some volatile declared variable and in the normal plugin calls you can then copy those values (and also apply the factor) and store it in the UserVar object.

clumsy-stefan commented:
Of course, you're right... didn't think of this... I'll adapt it accordingly and try it... so it would be backwards compatible.. thanks!!

clumsy-stefan committed:
P003 µSec backward compatibility

Make P003 backward compatible when using micros(). With this all values stay im miliseconds ("Time" with fractions)

clumsy-stefan commented:
A few test with the latest change show, that with this RPM's up to about 10'000 can be measured more or less accurately (~6ms cycle time). I tested with CPU-frequency of 80 and 160MHz, which showed the same results. But at least it's much more exect than with `milis()`

Count measured over a period of time and RPM calculated from that gives more or less the same result than RPM calculated from the "instant" "Time" value.

TD-er commented:
For such high interrupt counts I am really surprised the values aren't all over the range of possible values.
Good to know it is indeed possible to measure those short intervals.

I guess this may give very different results when running a plugin that may disable interrupts, like DHT plugins.
Or affecting stability of reading timing critical sensors like the Dallas sensors.

clumsy-stefan commented:
> I guess this may give very different results when running a plugin that may disable interrupts, like DHT plugins.
> Or affecting stability of reading timing critical sensors like the Dallas sensors.

Agree, but this is independent if you use `micros()` or `milis()`, the number of interrupts don't change.

I think this could be a neat improvment as it's possible to calculate more accurate "instant" values of RPM's especially in the range of 4'000-8'000 where the ms is just too unprecise.

TD-er commented:
Sure I agree on the improved resolution.
So I don't have any objections to the changes.

TD-er commented on src/_P003_Pulse.ino:
```diff
@@ -277,14 +279,14 @@ void Plugin_003_pulsecheck(byte Index)
   //  Alternatively there is timePassedSince(Plugin_003_pulseTimePrevious[Index]); but this is not in IRAM at this time, so do not use in a
   // ISR!
-  const unsigned long PulseTime = millis() - Plugin_003_pulseTimePrevious[Index];
+  const unsigned long PulseTime = micros() - Plugin_003_pulseTimePrevious[Index];
```
This will cause strange issues when the `micros()` will overflow.
Was already an issue with `millis()` but that would only happen every 49.7 days.
`micros()` will overflow roughly 0.0497 days :) (71 minutes)
So that will happen quite often.

You should use the `timePassedSince` call for this.
Or maybe use `micros64()` and store `Plugin_003_pulseTimePrevious` and `Plugin_003_pulseTime` should be stored as `uint64_t`

clumsy-stefan replied:
Hmm.. I am running this for over 2 days now on a node with 2 pulse input's at roughly 2000RPM and didn't have any issue so far. I guess what happens is, that you will just "loose" that one event when the counter overflows.... which is not nice, but not really a problem I think...
the bigger problem is, that you do not want to do too much in this routine as it's time critical...
I'm happy to look into it, but as said, I don't thik that it's a real problem...

TD-er replied:
At high pulse rates it is not likely to be a problem, but the pulse plugin is also used to measure pulses which may be more than an hour apart.
For example water consumption counter.

clumsy-stefan replied:
latest commit uses `uint64_t micros64()` internally

clumsy-stefan committed:
P003 use micros64()

Use micros64() instead of micros() to increase the overflow time. Due to the conversions the overflow will now happen roughly every 80'000 minutes.

TD-er commented on src/_P003_Pulse.ino:
```diff
@@ -194,7 +194,7 @@ boolean Plugin_003(byte function, struct EventStruct *event, String& string)
           UserVar[event->BaseVarIndex]     = Plugin_003_pulseCounter[event->TaskIndex];
           UserVar[event->BaseVarIndex + 1] = Plugin_003_pulseTotalCounter[event->TaskIndex];
-          UserVar[event->BaseVarIndex + 2] = Plugin_003_pulseTime[event->TaskIndex]/1000.0f;
+          UserVar[event->BaseVarIndex + 2] = (unsigned long)(Plugin_003_pulseTime[event->TaskIndex]/1000.0f);
```
Why cast it to unsigned long?
You wanted to increase resolution, but now you cast it to msec again.

clumsy-stefan replied:
That's only the UserVar (which I assumed is 32bit) that stores the PulseTime (in ms anyways, for compatibility reasons,, therefore `/1000.0f`), so there the overflow will (still) happen as before on the `milis()` version (some 50days).

But for the internal calculation of the difference 64bit values are used everywhere and calculations are don in µs.

That's just IMHO. though...

TD-er replied:
You convert it to msec by dividing with 1000.0
but if you cast it then to an integer type, you will loose the gained resolution, so there is no extra resolution anymore from when we started this PR.

`millis()` (e.g. 123 ms) -> `micros() / 1000.0` (e.g. 123.456 ms) -> cast to int -> millis resolution again. (e.g. 123 ms)

Sure by using a 64 bit counter the 49.7 days bug is now fixed, but you're throwing away the microsec resolution.

clumsy-stefan replied:
of course you're right (as always ;) this probably explains the strange issue I had (in counting pulses).... this should be casted to a float I guess, which was my intention but I just messed up... will correct it!

TD-er replied:
When you assign it to the UserVar, it will automatically be casted to float, as it is of type float.

clumsy-stefan replied:
ok, thanks. So you would suggest to skip it? could leave it to make it clear, that these are different types, but it's your project ;) I'll adhere to your guidelines...
EDIt: already did a new version with a cast to float, I'll see if it stays stable..)

clumsy-stefan replied:
done. also added `getMicros()` in commit #61e05159f9a036a9568e80bdb8a3964443044cbd

TD-er commented:
Can you change the call to `micros64()` into `getMicros64()` ?
You may need to `#include "src/Helpers/ESPEasy_time_calc.h"`
That's a wrapper I made for both ESP8266 and ESP32 as they use different calls for getting the 64 bit microsec counter.

clumsy-stefan commented:
Sure, will that still work in the interrupt handling?

there is anyways somtehing failing, as I get restarts after more or less exactly 5min. now from a HW watchdog even though, there is no PWM signal attached currently (but the plugin is active). But as I'm in the office currently and the test-fan is in my living room, I cant really debug what's happening... My wife always gets a heart-attack when I increase the fan-speed to 10'000RPM ;)

I'll include toe suggested getMicros and see what happens..

TD-er commented:
Well you can change it back to millis() to test if it still triggers the HW watchdog.
But I doubt it is related.
Watchdog usually occurs on blocking code, which can be seen (before the reboot :) ) in the timing stats.

clumsy-stefan commented:
There has to be something much more fundamental, I have a "empty" testunit (see pic. below) that just rebooted with:
```
Reset Reason: | Hardware Watchdog
Last Action before Reboot: | Const Interval: TIMER_20MSEC
```
I still have no clue what's going on, so I'll debug further.... but I don't think it^s related to your time-fix PR... But as here also no Pulse is active, I also think it^s not related to that plugin.... 

![image](https://user-images.githubusercontent.com/32835100/107648209-4daf2300-6c7c-11eb-9c9b-78e57972f551.png)

TD-er commented:
What controller do you have active?
HW watchdog is quite often related to a network disconnect which remains unnoticed.

clumsy-stefan commented:
I have only C009 (FHEM) included in the build and active... But yes, sometimes the server doesn't answer (although the network is connected)... but a few days ago I never had this... my build from the weekend runs on 30 nodes without rebooting... this happens this often only since the last merge.... but I'm currently clueless why...

EDIT: I even have a build from yesterday where it doesn't happen... I have to find out what chnged since then...

TD-er commented:
You are using the latest esp8266/Arduino lib?
I am now testing with that one too and also see some weird crashes (not WDT reboots) so maybe something there was also changed?

clumsy-stefan commented:
Yes I do, but there nothing changed between yesterday and today I think... What's really strange is, that it always happens exactly after a multiple of 5 minutes. I have 4 untis running this code now and they fail either at 5, 10, 20 or 40 min. uptime... 
I'm trying to debug a bit more, I have two of these units serial attached, probably I can capture a bit more what's going on...

TD-er commented:
Maybe an NTP update?
Or trying to resolve a DNS hostname which has a multiple of 5 minutes as TTL?
Not sure though whether the DNS resolve always obeys the TTL set in the DNS reply.
I do remember to have seen some issues where it may use its own TTL in LWIP stack.

clumsy-stefan commented:
Ok, so just a wild guess: the only thing I see that changed between yesterday'ss build where I used the PR and the other builds where I used the mega branch is commit #082c02b5e47e80d73ccdb812b847c4ad0b8e7497 ([WiFi] Fix crash when trying to reconnect while low on memory).

I did now a new version, with this commit reverted and insstalled it on the 4 test nodes (all with different configurations/sensors attached). Until now (~20min.) no more WD/reboots....

EDIT: what I see in timing stats are relatively high `WiFi.isConnected() (fail) 755	0.60	0.13	1.025	2.232	891.268` numbers...

TD-er commented:
It makes sense if that commit causes a reboot after 5 minutes, but not later.
Unless your units loose connection after 10 minutes and reconnect.

clumsy-stefan commented:
I use CPU-Eco Mode so I guess it could be (because of this or other things happening) that the unit dis- and reconnects to the AP from time to time, but unfortunately I didn't look on the number of reconnects. But if you want I can reflash the previous version and see if there is any relation between reconnects and reboots..

WIthout this commit the units run for nearly an hour now... so Iguess it's related to the changes there...

TD-er commented:
The reason for this commit was because of crashes when the AP candicate created from RTC is removed from the candidates list.

clumsy-stefan commented:
ok, but it definitely seems to cause the WD reboots... sorry for that...

TD-er commented:
Can you test this alternative way of handling WiFi AP candidate from RTC?
https://github.com/letscontrolit/ESPEasy/pull/3507

N.B. the other change in this commit is something I'm not entirely sure about, but given the crash stack decode I got, I made sure to change it.

```
--------------- CUT HERE FOR EXCEPTION DECODER ---------------

Exception (28):
epc1=0x4028ff74 epc2=0x00000000 epc3=0x00000000 excvaddr=0x0e520000 depc=0x00000000

LoadProhibited: A load referenced a page mapped with an attribute that does not permit loads
  epc1=0x4028ff74 in strnlen_P at /workdir/repo/newlib/newlib/libc/sys/xtensa/string_pgmspace.c:51

>>>stack>>>

ctx: cont
sp: 3fff2760 end: 3fff2a50 offset: 0190
3fff28f0:  402d81c0 00000002 3fff293c 4027b15c  
3fff2900:  00000001 0e520000 3fff2930 4027b40f
3fff2910:  00000004 0e520000 3fff2970 4027b441
3fff2920:  00000004 402d8208 3fff2970 4023caed
3fff2930:  00000000 3fff9db4 00000000 0000232f
3fff2940:  3fff9e82 02000000 3fff2970 3fff9db4
3fff2950:  00000001 3fff9db4 3fff2970 4023cc5c
3fff2960:  3fff9db4 00000000 00000001 4023d3f3
3fff2970:  3fff93ac 0010001f 80000000 00000000
3fff2980:  3ffeff34 00000000 00000000 40251a0c
3fff2990:  00000000 3fff9db4 3fff562c 40211bf6
3fff29a0:  3ffeff34 3fff01c8 0000000a 3fff01c8
3fff29b0:  3ffeff34 00000001 3fff0638 40251a51
3fff29c0:  002a26de 3fff01c8 40253d1c 40253d96
3fff29d0:  000009cd 002a2ce5 40254834 40254854
3fff29e0:  00000000 000008b5 3fff0100 40242988  
3fff29f0:  00000004 3fff16b8 0003d712 3fff2a9c
3fff2a00:  3fffdad0 3fff188c 00000000 3fff2a9c
3fff2a10:  3fffdad0 3ffe8678 00000000 40230c0d
3fff2a20:  00000000 3fff188c 00000000 40100795
3fff2a30:  3fffdad0 3fff188c 3fff2a80 4027c7e4
3fff2a40:  feefeffe feefeffe 3ffe87a4 401006d1
<<<stack<<<

0x402d81c0 in chip_v6_unset_chanfreq at ??:?
0x4027b15c in String::copy(__FlashStringHelper const*, unsigned int) at ??:?
0x4027b40f in String::operator=(__FlashStringHelper const*) at ??:?
0x4027b441 in String::String(__FlashStringHelper const*) at ??:?
0x402d8208 in chip_v6_unset_chanfreq at ??:?
0x4023caed in String::replace(__FlashStringHelper const*, __FlashStringHelper const*) at ??:?
0x4023cc5c in getLWT_topic(ControllerSettingsStruct const&) at ??:?
0x4023d3f3 in MQTTCheck(unsigned char) at ??:?
0x40251a0c in updateMQTTclient_connected() at ??:?
0x40211bf6 in PubSubClient::loop() at ??:?
0x40251a51 in runPeriodicalMQTT() at ??:?
0x40253d1c in ESPEasy_Scheduler::process_interval_timer(ESPEasy_Scheduler::IntervalTimer_e, unsigned long) at ??:?
0x40253d96 in ESPEasy_Scheduler::process_interval_timer(ESPEasy_Scheduler::IntervalTimer_e, unsigned long) at ??:?
0x40254834 in ESPEasy_Scheduler::handle_schedule() at ??:?
0x40254854 in ESPEasy_Scheduler::handle_schedule() at ??:?
0x40242988 in handle_unprocessedWiFiEvents() at ??:?
0x40230c0d in loop at ??:?
0x40100795 in esp_schedule at ??:?
0x4027c7e4 in loop_wrapper() at core_esp8266_main.cpp:?
0x401006d1 in cont_wrapper at ??:?
```

clumsy-stefan committed:
P003 use getMicros64() function

To make sure this works on esp32 as well as on esp8266 use the helper function getMicros64() instead of using micros64() directly.

clumsy-stefan commented:
> Can you test this alternative way of handling WiFi AP candidate from RTC?
> #3507

Sure I can. Probably we should continue the discussion in the relevant PR and leave this to the P003 change.

I just pushed the commit for `getMicros64()` as suggested. I got rid now of the "strange" readings and counts I got... a (typical) beginner mistake. I used "normal" cable (~1m length) to connect the fan to the ESP, wih at some combinations of PWM and RPM signal gave random readings... After changing it to a grounded twisted cable (PWM/GND and RPM/GND) readings up to ~10'000 RPM were consistant between number of Pulses and Time between pulses (with some allowance)... So I also added a comment about that in the Plugin source. Probably this should be written as a comment on the webpage at some point in time...

clumsy-stefan committed:
Merge branch 'mega' into P003_Pulse_Micros

clumsy-stefan commented:
@TD-er are you still requiring any changes here or any reservations on the changes?

TD-er commented:
Did you test the code for a while?
Mainly to make sure we don't run into "ISR" related issues where code is used from callback functions not being flagged to be located in IRAM.

clumsy-stefan commented:
Sure, have it running for roughly a week now on all my nodes. One of them (the heavy loaded one with 24pcfgpio's and other sensors attached, see screenshot) has two PWM fan's attached with 2 pulse-counters that send between 1500 and 7500 PPM. Tested it up to about 12'000RPM on a test node, also with no visible side effects.

![image](https://user-images.githubusercontent.com/32835100/108346407-9ff1c600-71df-11eb-9faa-4bf19cd50e74.png)
(RPM1/2 are the FAN's)

EDIT:
PS: the ISR code did not change at all! all calculations are done outside the ISR routines.

TD-er commented:
> PS: the ISR code did not change at all! all calculations are done outside the ISR routines.

But you're calling the `getMicros64()` function from the ISR.
Still good to know it is now running stable with quite high frequencies to count.

clumsy-stefan commented:
> But you're calling the getMicros64() function from the ISR.

Right... but it seems to be no issue at all... I' quite confident that this PR only improves the `Time` accuracy without any othe implications.

TD-er commented:
Me too, so therefore I merged it :)
